### PR TITLE
Feature: Generic discovery and poller tests

### PR DESCRIPTION
**Pull request by murrant**
DO NOT DELETE THIS TEXT

#### Please note

> Please read this information carefully. You can run `./scripts/pre-commit.php` to check your code before submitting.

- [x] Have you followed our [code guidelines?](http://docs.librenms.org/Developing/Code-Guidelines/)

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`

### CODE REVIEW

File: scripts/save-test-data.php [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+        'module:',
+        'os:',
+        'variant:',
+        'file:',
+    )
+);
+
+$debug = (isset($options['d']) || isset($options['debug']));
+$vdebug = $debug;
+$snmpsim_ip = '127.1.6.1';
```
Comment (laf) [inline scripts/save-test-data.php]: I've not come across 127.1.6.1 used as a localhost IP before, is it sensible to use it over 127.0.0.1?

**Comment by murrant** [comment]
Yeah, requiring testing is a big barrier, I am worried about that too, but I hope to make it dead simple with the script.  The benefits of having the tests should be pretty clear :)

`snmpsim.error.SnmpsimError: broken record <b'1.3.6.1.2.1.25.3.2.1.3.196608||Intel Xeon E3-12xx v2 (Ivy Bridge)\n'>`

I fixed a lot of issues with the save-test-data.php script today, remove the snmprec file and try again.

**Comment by laf** [comment]
Works better now.

For me running `./scripts/save-test-data.php -d --no-save -h localhost -o linux -m processors` it just changes some of the existing data in the json and snmprec, wouldn't this just mean we're constantly changing the data within the file based on people's local setup?

**Comment by murrant** [comment]
Yes, but that is why we should use variants.  Once all the modules are added. We won't need to update specific os captures much.

Also for devices, the data won't change as wildly as it does for servers.

When merging existing snmprec and new snmprec I could prefer existing data over new?  Which do you think is better?

**Comment by laf** [comment]
> When merging existing snmprec and new snmprec I could prefer existing data over new? Which do you think is better?

Seems better if the OIDs exist in the test data not to overwrite them.

**Comment by laf** [comment]
If you can add docs for this, after the overriding of data is fixed then this is good for a merge.

**Comment by murrant** [comment]
Hmm, for some modules we will need separate data recorded for discovery and polling :/

**Comment by laf** [comment]
I've still not read the code so some of this will be answered in there I'm sure.

`./scripts/save-test-data.php -h localhost -o linux -d`

Still adds data to the test files, is that expected:

```
-bash-4.2$ git diff tests/data/linux.json
diff --git a/tests/data/linux.json b/tests/data/linux.json
index 8607a2d7d..4e8107cc1 100644
--- a/tests/data/linux.json
+++ b/tests/data/linux.json
@@ -93,5 +93,9 @@
             ]
         },
         "poller": "matches discovery"
+    },
+    "all": {
+        "discovery": [],
+        "poller": "matches discovery"
     }
 }
\ No newline at end of file
-bash-4.2$ git diff tests/snmpsim/linux.snmprec
diff --git a/tests/snmpsim/linux.snmprec b/tests/snmpsim/linux.snmprec
index fcbd58141..36a074e59 100644
--- a/tests/snmpsim/linux.snmprec
+++ b/tests/snmpsim/linux.snmprec
@@ -1,5 +1,10 @@
 1.3.6.1.2.1.1.1.0|4|Linux some.host.name 4.9.0 #1 SMP Mon Dec 12 14:44:53 CST 2016 x86_64
 1.3.6.1.2.1.1.2.0|6|1.3.6.1.4.1.8072.3.2.10
+1.3.6.1.2.1.1.3.0|67|24144912
+1.3.6.1.2.1.1.4.0|4|<private>
+1.3.6.1.2.1.1.5.0|4|<private>
+1.3.6.1.2.1.1.6.0|4|<private>
+1.3.6.1.2.1.25.1.1.0|67|24146459
 1.3.6.1.2.1.25.3.2.1.1.196608|2|196608
 1.3.6.1.2.1.25.3.2.1.1.196609|2|196609
 1.3.6.1.2.1.25.3.2.1.1.196610|2|196610
@@ -167,3 +172,4 @@
 1.3.6.1.2.1.25.3.8.1.9.28|4|0-1-1,0:0:0.0
 1.3.6.1.2.1.25.3.8.1.9.29|4|0-1-1,0:0:0.0
 1.3.6.1.2.1.25.3.8.1.9.30|4|0-1-1,0:0:0.0
+1.3.6.1.6.3.10.2.1.3.0|2|241449
```

I do find that I need to run the script twice, the first time snmpsimd doesn't start the first time for some reason.

**Comment by murrant** [comment]
I have not added support for more than one module yet, but  I plan to.  So I left the option to do all in the script.

I just changed the snmpsim stuff around a little I'll give it some more testing, but I don't remember that happening for me.

**Comment by laf** [comment]
You happy with this to be merged then @murrant? I'm happy that it can just be adjusted as we go along if needs be?

**Comment by murrant** [comment]
Yeah, I think the file format is good.  Just needs expansion and bug fixes now.

**Comment by laf** [comment]
> Yeah, I think the file format is good. Just needs expansion and bug fixes now.

Still not sure if that means it's ready for a merge @murrant :)

I'm trying to fix the travis issue at the moment, think I'm going to resort to just dropping the check for poller-service.py for now.

**Comment by laf** [comment]
Travis is now fixed so feel free to just rebase this and merge if you think it's ready.

**Pull request merged**